src/gateway/energy/energy_bus_controller.py

```diff
@@ -0,0 +1,451 @@
+            for port_id in range(EnergyEnums.NUMBER_OF_PORTS[module_version]):
+                ct = EnergyCT(number=port_id,
+                              sensor_type=2,  # Default of 12.5A
+                              times='',  # No times by default (all night)
+                              energy_module=energy_module)
+                db.add(ct)
+            db.commit()
+
+    @staticmethod
+    def _get_next_unused_address():
+        with Database.get_session() as db:
+            modules = db.query(Module).where(Module.source == ModuleDTO.Source.GATEWAY,
+                                             Module.hardware_type == HardwareType.PHYSICAL).all()
+            used_addresses = [x.address for x in modules]
+            new_address = 2
+
+            while new_address in used_addresses:
+                new_address += 1
+
+            return new_address
+
+    @staticmethod
+    def _find_address_for_serial(serial_number):
+        # type: (str) -> Tuple[int, bool]
+        with Database.get_session() as db:
+            module = db.query(Module).where(Module.source == ModuleDTO.Source.GATEWAY,
+                                            Module.hardware_type == HardwareType.PHYSICAL,
+                                            Module.serial_number == serial_number).first()
+            if module is None:
+                address = EnergyBusDiscoveryMode._get_next_unused_address()
+            else:
+                address = int(module.address)
+
+            return address, module is not None
+
+    @staticmethod
+    def receive_packet(context, energy_packet):
+        # type: (EnergyBusContext, EnergyPacket) -> None
+        logger.debug("Received an packet in discovery mode: {}".format(energy_packet))
+        if energy_packet.type not in ['ADD']:
+            logger.error('Energy Bus: Received unexpected packet in addressing mode: {}'.format(energy_packet.printable_data()))
+            return
+        if energy_packet.mode != 'G':
+            logger.error('Energy Bus: Received Want An Address packet, but not with the right mode: {}'.format(energy_packet.printable_data()))
+            return
+
+        cmd = EnergyAPI.get_new_address(EnergyEnums.Version.ENERGY_MODULE)
+        cmd.reply = energy_packet
+        serial_number, random_seed = cmd.read_output()
+        serial_number = serial_number.decode()
+        module_version = energy_packet.address_type
+
+        if not isinstance(serial_number, str):
+            logger.error('Received Get ADDress packet, but no valid data in packet: {}'.format(energy_packet))
+            return
+        if module_version is None:
+            logger.error('Received Get ADDress packet, address type is None: {}'.format(energy_packet))
+            return
+
+        logger.debug("Received discovery packet from energy bus from serial: {}; {}".format(serial_number, random_seed))
+        data = (random_seed, module_version)
+        if serial_number in context.discovery_buffer:
```

Comment:
This is a bit more concise.

```suggestion
        context.discovery_buffer.setdefault(serial_number, [])
        context.discovery_buffer[serial_number].append(data)
```